src/jltypes.c

```diff
@@ -2844,6 +2867,15 @@ jl_vararg_t *jl_wrap_vararg(jl_value_t *t, jl_value_t *n, int check, int nothrow
     return vm;
 }
 
+// compute a conservative estimate of whether there could exist an instance of a subtype of this
+void jl_compute_has_concrete_subtype_from_fields(jl_datatype_t *dt) {
+    size_t nfields = jl_svec_len(dt->types);
+    for (size_t i = 0; dt->has_concrete_subtype && i < nfields - dt->name->n_uninitialized; i++) {
+        jl_value_t *fld = jl_svecref(dt->types, i);
+        dt->has_concrete_subtype = jl_has_concrete_subtype(fld);
```

Comment:
Oddly enough, I don't think you necessarily want `jl_has_concrete_subtype`, but rather `jl_precompute_has_concrete_subtype`. The former is partially recursive (because it examines has_concrete_subtype), which makes its result somewhat unstable